packages/store-sync/src/recs/syncToRecs.ts

```diff
@@ -0,0 +1,227 @@
+  components: TComponents;
+  initialState?: {
+    blockNumber: bigint | null;
+    tables: (Table & { records: TableRecord[] })[];
+  };
+};
+
+type SyncToRecsResult<
+  TConfig extends StoreConfig = StoreConfig,
+  TComponents extends Record<string, RecsComponent<RecsSchema, StoreComponentMetadata>> = Record<
+    string,
+    RecsComponent<RecsSchema, StoreComponentMetadata>
+  >
+> = {
+  // TODO: return publicClient?
+  components: TComponents & ReturnType<typeof defineInternalComponents>;
+  singletonEntity: Entity;
+  latestBlock$: Observable<Block>;
+  latestBlockNumber$: Observable<bigint>;
+  blockLogs$: Observable<BlockLogs>;
+  blockStorageOperations$: Observable<BlockStorageOperations<TConfig>>;
+  waitForTransaction: (tx: Hex) => Promise<{ receipt: TransactionReceipt }>;
+  destroy: () => void;
+};
+
+export async function syncToRecs<
+  TConfig extends StoreConfig = StoreConfig,
+  TComponents extends Record<string, RecsComponent<RecsSchema, StoreComponentMetadata>> = Record<
+    string,
+    RecsComponent<RecsSchema, StoreComponentMetadata>
+  >
+>({
+  world,
+  config,
+  address,
+  publicClient,
+  components: initialComponents,
+  initialState,
+}: SyncToRecsOptions<TConfig, TComponents>): Promise<SyncToRecsResult<TConfig, TComponents>> {
+  const components = {
+    ...initialComponents,
+    ...defineInternalComponents(world),
+  };
+
+  const singletonEntity = world.registerEntity({ id: hexKeyTupleToEntity([]) });
+
+  let startBlock = 0n;
+
+  if (initialState != null && initialState.blockNumber != null) {
+    debug("hydrating from initial state to block", initialState.blockNumber);
+    startBlock = initialState.blockNumber + 1n;
+
+    setComponent(components.SyncProgress, singletonEntity, {
+      step: SyncStep.SNAPSHOT,
+      message: `Hydrating from snapshot to block ${initialState.blockNumber}`,
+      percentage: 0,
+    });
+
+    const componentList = Object.values(components);
+
+    const numRecords = initialState.tables.reduce((sum, table) => sum + table.records.length, 0);
+    let recordsProcessed = 0;
+
+    for (const table of initialState.tables) {
+      setComponent(components.TableMetadata, getTableKey(table) as Entity, { table });
+      const component = componentList.find((component) => component.id === table.tableId);
+      if (component == null) {
+        debug(`no component found for table ${table.namespace}:${table.name}, skipping initial state`);
+        continue;
+      }
+      for (const record of table.records) {
+        const entity = encodeEntity(table.keySchema, record.key);
+        setComponent(component, entity, record.value as ComponentValue);
+
+        recordsProcessed++;
+        setComponent(components.SyncProgress, singletonEntity, {
+          step: SyncStep.SNAPSHOT,
+          message: `Hydrating from snapshot to block ${initialState.blockNumber}`,
+          percentage: (recordsProcessed / numRecords) * 100,
+        });
```

Comment:
I think we should only update the sync progress every x records (eg 100 or 1000), otherwise every other update event is a progress update event, which might unnecessarily overwhelm the client if there are a lot of records in the initial state (think OPCraft)

Reply:
I did a naive thing for now, but would like to move a bunch of the sync stuff over to a less-blocking approach with `requestIdleCallback` or similar.